### PR TITLE
prov/efa: Initiate peer_srx struct and ops

### PR DESCRIPTION
**Pull request by shijin-aws**
This PR add peer_srx to rxr_ep, and initiate its fields and ops functions as stubs. This is a preliminary step that allows future incremental changes.

### CODE REVIEW

File: prov/efa/src/rdm/rxr_ep.c [hunk 1]
```diff
@@ -2189,6 +2190,8 @@ int rxr_endpoint(struct fid_domain *domain, struct fi_info *info,
 	if (ret)
 		goto err_free_ep;
 
+	rxr_peer_srx_construct(&rxr_ep->peer_srx);
```
Comment (shijin-aws) [inline prov/efa/src/rdm/rxr_ep.c:2193]: I can drop this change if requested. But I think this will not have any impact on the current code.

**Comment by sunkuamzn** [comment]
Could we change the prefix for file name, structs and functions from `rxr` to `efa_rdm`? We are trying to move to `efa_rdm`

**Comment by shijin-aws** [comment]
> Could we change the prefix for file name, structs and functions from `rxr` to `efa_rdm`? We are trying to move to `efa_rdm`

sure. will do

File: prov/efa/src/rdm/rxr_ep.h [hunk 2]
```diff
@@ -234,6 +234,8 @@ struct rxr_ep {
 	int blocking_copy_rx_entry_num; /* number of RX entries that are using gdrcopy/cudaMemcpy */
 
 	int	hmem_p2p_opt; /* what to do for hmem transfers */
+
+	struct fid_peer_srx peer_srx; /* support sharing receive context with peer providers */
```
Comment (wenduwan) [inline prov/efa/src/rdm/rxr_ep.h:238]: Suggest moving this to L74 because of the its significance and relevance to SHM

File: prov/efa/src/rdm/rxr_srx.c [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "efa.h"
+#include "rxr_srx.h"
```
Comment (wenduwan) [inline prov/efa/src/rdm/rxr_srx.c:35]: Question - is it possible to move this to `prov/efa/src`?

Asking because I also see `prov/efa/src/efa_shm.h` which is relevant to this who peer API thing.

File: prov/efa/src/rdm/rxr_srx.c [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+/**
+ * @brief This call is invoked by the peer provider to obtain the receive
+ * buffers where an incoming message should be placed.
+ *
+ * @param[in] srx the fid_peer_srx
+ * @param[in] addr the source address of the incoming message
+ * @param[in] size the size of the incoming message (ignored unless multi-receive)
+ * @param[out] rx_entry the obtained receive buffer
```
Comment (wenduwan) [inline prov/efa/src/rdm/rxr_srx.c:44]: Based on the signature `struct fi_peer_rx_entry **rx_entry` I assume this is a list of buffers instead of a singular buffer?

File: prov/efa/src/rdm/rxr_srx.c (hunk 3 above)
Comment (shijin-aws) [reply prov/efa/src/rdm/rxr_srx.c:35]: I do not think efa will use shm for dgram, so I will keep these files inside `rdm`. We probably need to move the `efa_shm.h` to `rdm/`

File: prov/efa/src/rdm/rxr_srx.c (hunk 4 above)
Comment (shijin-aws) [reply prov/efa/src/rdm/rxr_srx.c:44]: No, it's a single buffer, it's `**` because this function need to allocate a new `fi_peer_rx_entry`.

Comment (wenduwan) [reply prov/efa/src/rdm/rxr_srx.c:44]: I see. Then maybe rephrase the description:
`This call is invoked by the peer provider to obtain the receive buffers` -> `... buffer`

File: prov/efa/src/rdm/rxr_srx.c [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+ * @brief This call is invoked by the peer provider to obtain the receive
+ * buffers where an incoming tagged message should be placed.
+ *
+ * @param[in] srx the fid_peer_srx
+ * @param[in] addr the source address of the incoming message
+ * @param[in] size the size of the incoming message (ignored unless multi-receive)
+ * @param[out] rx_entry the obtained receive buffer
+ * @return int 0 on success, a negative integer on failure
+ */
+static int rxr_srx_get_tag(struct fid_peer_srx *srx, fi_addr_t addr,
```
Comment (wenduwan) [inline prov/efa/src/rdm/rxr_srx.c:63]: `*_tag` is ambiguous IMO, suggest being a verbose name, i.e. `*_tagged_msg`

File: prov/efa/src/rdm/rxr_srx.c [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+static void rxr_srx_free_entry(struct fi_peer_rx_entry *entry)
+{
+}
+
+/**
+ * @brief This call is invoked by the owner provider to start progressing
+ * the rx_entry that matches a received message.
+ *
+ * @param rx_entry the rx entry to be progressed.
+ * @return int 0 on success, a negative integer on failure.
```
Comment (wenduwan) [inline prov/efa/src/rdm/rxr_srx.c:112]: Comment consistency:
```
@param[in] rx_entry the rx entry to be progressed.
```

File: prov/efa/src/rdm/rxr_srx.c [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+ */
+static int rxr_srx_start_msg(struct fi_peer_rx_entry *rx_entry)
```
Comment (wenduwan) [inline prov/efa/src/rdm/rxr_srx.c:114]: Do you think replacing `start` with `progress` better describes the function?

File: prov/efa/src/rdm/rxr_srx.c [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+	.free_entry = rxr_srx_free_entry,
+};
+
+static struct fi_ops_srx_peer rxr_srx_peer_ops = {
+	.size = sizeof(struct fi_ops_srx_peer),
+	.start_msg = rxr_srx_start_msg,
+	.start_tag = rxr_srx_start_tag,
+	.discard_msg = rxr_srx_discard_msg,
+	.discard_tag = rxr_srx_discard_tag,
+};
```
Comment (wenduwan) [inline prov/efa/src/rdm/rxr_srx.c:175]: Suggest moving to L106 to denote the end of peer ops

File: prov/efa/src/rdm/rxr_srx.c (hunk 7 above)
Comment (shijin-aws) [reply prov/efa/src/rdm/rxr_srx.c:114]: I prefer to keep it as `start` to be consistent with the ops name `start_msg` in the public API.

File: prov/efa/src/rdm/rxr_srx.c (hunk 5 above)
Comment (shijin-aws) [reply prov/efa/src/rdm/rxr_srx.c:63]: I may change it to `*_tag_entry` to be consistent with the API name. But I agree with you that this API name is kind of not perfect.

File: prov/efa/src/rdm/rxr_ep.h (hunk 2 above)
Comment (shijin-aws) [reply prov/efa/src/rdm/rxr_ep.h:238]: Conceptually this `peer` can be any provider, not for shm specifically. I do not think making it close to `shm` variables make more sense.

File: prov/efa/src/rdm/rxr_srx.c (hunk 5 above)
Comment (shijin-aws) [reply prov/efa/src/rdm/rxr_srx.c:63]: No, the man page is wrong ... the API name is exactly `get_tag`. I will keep the current `get_tag` to be consistent.

File: prov/efa/src/rdm/rxr_srx.c (hunk 8 above)
Comment (shijin-aws) [reply prov/efa/src/rdm/rxr_srx.c:175]: I guess you are saying moving `rxr_srx_owner_ops` to L106. I prefer to keep it in this way so the ops function table assignments can be put together.

File: prov/efa/src/rdm/rxr_srx.c (hunk 4 above)
Comment (shijin-aws) [reply prov/efa/src/rdm/rxr_srx.c:44]: Makes sense. I further rephrase it as `the receive buffers-> a rx_entry`

**Review by sunkuamzn — APPROVED** [review]

**Comment by sunkuamzn** [comment]
> We can start from making fi_peer_rx_entry as a new member in rxr_op_entry , implement the searching procedure with rxr_op_entry , then output rxr_op_entry->fi_peer_rx_entry .

Do you want to add fi_peer_rx_entry to rxr_op_entry in this PR or a separate one later? Both owner ops and peer ops need that member.

**Comment by shijin-aws** [comment]
>Do you want to add fi_peer_rx_entry to rxr_op_entry in this PR or a separate one later? Both owner ops and peer ops need that member.

Good point, I will add it in this PR, thanks!

**Review by wenduwan — APPROVED** [review]

**Review by sunkuamzn — APPROVED** [review]

**Comment by shijin-aws** [comment]
This PR may still need some change, before resolving my conversation with @aingerson  in https://github.com/ofiwg/libfabric/pull/8456#discussion_r1131653686.

**Comment by shijin-aws** [comment]
Updated the PR to set `peer_srx->ep_fid.fid.context = rxr_ep` so the peer_srx ops function can access owner resources in `rxr_ep` through `peer_srx->ep_fid.fid.context`

File: prov/efa/src/rdm/efa_rdm_srx.c [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+ * @brief This call is invoked by the peer provider to obtain a
+ * rx_entry where an incoming message should be placed.
+ * @param[in] srx the fid_peer_srx
+ * @param[in] addr the source address of the incoming message
+ * @param[in] size the size of the incoming message
+ * @param[out] rx_entry the obtained rx_entry
+ * @return int 0 on success, a negative integer on failure
+ */
+static int efa_rdm_srx_get_msg(struct fid_peer_srx *srx, fi_addr_t addr,
+		       size_t size, struct fi_peer_rx_entry **rx_entry)
```
Comment (sunkuamzn) [inline prov/efa/src/rdm/efa_rdm_srx.c:47]: I think calling all `fi_peer_rx_entry` structs `peer_rx_entry` or `rx_peer_entry` would avoid confusion with `rxr_op_entry` structs. `rx_entry` is overloaded..

Comment (shijin-aws) [reply prov/efa/src/rdm/efa_rdm_srx.c]: Updated

**Review by sunkuamzn — APPROVED** [review]
